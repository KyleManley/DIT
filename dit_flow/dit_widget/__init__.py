import dit_widget.common
import dit_widget.mathfamily
import dit_widget.printfamily
import dit_widget.replacefamily

__all__ = ['add_const',
           # TODO: add break_columns
           'check_int',
           'count_records',
           'count_values',
           'create_gtnp_metadata_json',
           'decimal_to_minsec',
           'div_const',
           'find_tz',
           'latlong_to_utm',
           # TODO: add map_read
           'mid_month',
           'minsec_to_decimal',
           'move_text',
           'mult_const',
           'pdf',
           # TODO: add plot_locations
           'print_ge',
           'print_gt',
           'print_le',
           'print_lt',
           'print_max',
           'print_notin_rangex',
           'print_rangex',
           'reformat_dates_to_gtnp',
           'remove_chars',
           'remove_duplicate',
           'remove_null',
           'replace_eq',
           'replace_ge',
           'replace_gt',
           'replace_le',
           'replace_lt',
           'replace_notin_rangex',
           'replace_rangex',
           'replace_text',
           'rounding',
           'sort_by_columns',
           'statistics',
           'sub_const',
           'translate_codes',
           'utm_to_latlong',
           ]
