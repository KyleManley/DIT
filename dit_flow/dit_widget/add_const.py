#!/usr/bin/python

"""Adds a constant to all numeric values in a column file."""

import sys
import getopt

from .mathfamily import arithmetic as a

__all__ = ['add_const']


def add_const(infile, outfile, constant):
    # Adds constant to all values in infile and writes the result to
    # outfile.
    a.arithmetic(infile, outfile, constant, lambda x, y: x + y,
                 lambda x, y: False)


def parse_args(args):
    def help():
        print('add_const.py -i <input file> -o <output file> -n <number>')

    infile = None
    outfile = None
    constant = None

    options = ('i:o:n:', ['input', 'output', 'constant'])
    readoptions = zip(['-' + c for c in options[0] if c != ':'],
                      ['--' + o for o in options[1]])

    try:
        (vals, extras) = getopt.getopt(args, *options)
    except getopt.GetoptError as e:
        print(str(e))
        help()
        sys.exit(2)

    for (option, value) in vals:
        if (option in readoptions[0]):
            infile = value
        elif (option in readoptions[1]):
            outfile = value
        elif (option in readoptions[2]):
            constant = float(value)

    if (any(val is None for val in [infile, outfile, constant])):
        help()
        sys.exit(2)

    return infile, outfile, constant

#                 PERFORM FUNCTION USING COMMAND-LINE OPTIONS                 #
if (__name__ == '__main__'):
    args = parse_args(sys.argv[1:])

    add_const(*args)
