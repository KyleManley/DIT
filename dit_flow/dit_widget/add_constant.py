#!/usr/bin/python
"""Adds a constant to all numeric values in a column file."""
import csv

import rill

import common.definitions as d


@rill.component
@rill.inport('INFILE')
@rill.inport('OUTFILE_IN')
@rill.outport('OUTFILE_OUT')
@rill.inport('CONSTANT')
def add_constant(INFILE, OUTFILE_IN, OUTFILE_OUT, CONSTANT):
    # Adds constant to all values in infile and writes the result to
    # outfile.
    for infile, outfile, constant in zip(INFILE.iter_contents(),
                                         OUTFILE_IN.iter_contents(),
                                         CONSTANT.iter_contents()):
        with open(infile, newline='') as _in, \
             open(outfile, 'w', newline='') as _out:
            data = csv.reader(_in, )
            output = csv.writer(_out)
            for line in _in:
                for item in line:
                    if (float(item) not in d.missing_values):
                        value = float(item) + constant

        OUTFILE_OUT.send(outfile)
