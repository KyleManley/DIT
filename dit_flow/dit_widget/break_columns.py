"""LOOKING AHEAD: This module splits an input csv file into individual
columns for easier manipulation by widgets."""

import csv
import itertools


def shard(datafile, header=True, name_override=None):
    """
    Inputs:
        datafile:
        header: If True (the default), the first line of the file is a
            header. If False, there is only data in the file.
        name_override: If None (the default), names are auto-selected
            by looking at the column headers or their position within
            the file. If a sequence, items that are strings will replace
            the names of corresponding columns (the first item will be
            the name of the first column, etc...) and items that are
            None or nonexistent will allow the auto-selection methods
            to run.
    Outputs:
        Writes each column of the data to a different file. The names of
        these files will be either the column headers (if they exist) or
        their zero-indexed position in the file.

        For FILE=
        Station,Date,Temperature
        XXX123,1999-03-21,-4.30265

        For FILE2=
        XYZ987,1492-12-12,3.5

        shard(FILE) -> Station.column, Date.column, Temperature.column
        shard(FILE2, header=False) -> 0.column, 1.column, 2.column
        shard(FILE, name_override=('st', 'dt', 'temp')) ->
            st.column, dt.column, temp.column
        shard(FILE, name_override=(None, 'DateTime')) ->
            Station.column, DateTime.column, Temperature.column
    """
    with open(datafile, 'rb') as f:
        data = csv.reader(f)
        column_check(data)

        # Determine filenames
        firstline = data.readline()
        if (name_override is not None):
            # Use the names given instead of finding them
            # If there are too few names in name_override
            name = ['' for all in firstline]
            for i in range(len(firstline)):
                if (i < len(name_override) and name_override[i]):
                    # If there is a name given for this position...
                    name[i] = name_override[i]
                else:
                    if (header):
                        name[i] = firstline[i]
                    else:
                        name[i] = i
        elif (header):
            # Use the values found in the header as column file names
            name = firstline
        else:
            # Use column indices as names
            name = list(range(len(firstline)))
            f.seek(0)  # Return to the top of the file to read again

        # Open files
        files = [open(str(item) + '.column', 'w') for item in name]

        # Write columns to files
        try:
            for line in data:
                for (i, item) in enumerate(line):
                    files[i].write(str(item))
        except ValueError:
            close_all(files)

        # Close files before exiting
        close_all(files)


def column_check(data):
    # TODO: Check if this exhausts the iterator in containing scope
    """Expects a csv.reader object."""
    count = []
    for line in data:
        count.append(len(line))
    mean = round(float(sum(count)) / len(count))
    error = False
    for (i, item) in enumerate(count):
        if (item != mean):
            print('Line {0} has a different number of columns than the rest '\
                'of the file.'.format(i+1))
            error = True
    if (error):
        raise IndexError('One or more of the lines was flawed.')
    else:
        return True


def close_all(files):
    for opened in files:
        opened.close()


def reunite(outfile, *args):
    """Unites all column files into a single csv.
    Inputs:
        outfile: The name of the csv file to write into.
        *args: Any number of filenames which hold individual columns.
            The columns will be put into the output file in the order
            they appear in this function call.
    Outputs:
        Writes into outfile.
    """
    with open(outfile, 'wb') as f:
        final = csv.writer(f)
        try:
            for (i, name) in enumerate(args):
                files[i] = open(args[i])

            for line in zip(*files):
                final.writerow(line)
        finally:
            close_all(files)
