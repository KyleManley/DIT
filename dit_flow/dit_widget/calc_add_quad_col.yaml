type: manipulation_widget
description: Add two columns in quadrature (out = sqrt(column_a**2 +column_b**2))
input_1: missing_value
input_1_type: float
