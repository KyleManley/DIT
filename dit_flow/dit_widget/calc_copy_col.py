#!/usr/bin/python
"""Copy input column to output column"""
import argparse as ap
import csv

from dit_flow.dit_widget.common.setup_logger import setup_logger, DEFAULT_LOG_LEVEL


def calc_copy_col(input_data_file=None, output_data_file=None, log_file=None, log_level=DEFAULT_LOG_LEVEL):
    # output = in_column_A + in_column_B
    logger = setup_logger(__name__, log_file, log_level)
    assert input_data_file is not None, 'An input CSV file with columns of values.'
    assert output_data_file is not None, 'An output CSV file to write new values.'
    with open(input_data_file, newline='') as _in, \
            open(output_data_file, 'w', newline='') as _out:
            logger.info('Copy input column to output column')
            output = csv.writer(_out)
            reader = csv.reader(_in)
            for line in reader:
                output.writerow(line)


def parse_arguments():
    parser = ap.ArgumentParser(description='Copy input column to output column.')

    parser.add_argument('-i', '--input_data_file',
                        help='Step file containing input data to manipulate.')
    parser.add_argument('-o', '--output_data_file', help='Step file to store output data.')
    parser.add_argument('-l', '--log_file', help='Step file to collect log information.')

    return parser.parse_args()


if __name__ == '__main__':
    args = parse_arguments()

    calc_copy_col(args.input_data_file, args.output_data_file, args.log_file)
