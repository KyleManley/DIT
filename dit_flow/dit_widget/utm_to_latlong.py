#! /usr/bin/python

import csv

import utm
import argparse as ap

from dit_flow.dit_widget.common.setup_logger import setup_logger, DEFAULT_LOG_LEVEL


def utm_to_latlong(input_data_file=None, output_data_file=None, log_file=None, log_level=DEFAULT_LOG_LEVEL):
    """Converts UTM coordinates into latitude/longitude.
    assumes rows are easting, northing, zone number, either 'N' for northern
    hemisphere or 'S' for southern hemisphere
    """
    logger = setup_logger(__name__, log_file, log_level)

    # Check required input and output data file names were given.
    assert input_data_file is not None, 'An input CSV file with columns of values.'
    assert output_data_file is not None, 'An output CSV file to write new values.'

    _in = open(input_data_file, 'r')
    try:
        _out = open(output_data_file, 'w')
        try:
            data = csv.reader(_in)
            output = csv.writer(_out)
            for row_ind, row in enumerate(data):
                east = float(row[0])
                north = float(row[1])
                zone = int(row[2])

                latlong = utm.to_latlon(east, north, zone, northern=('N' == row[3]))
                logger.info('Changed row {} from: {}  to: {}'.format(row_ind,
                                                                     (row[0], row[1]), latlong))

                output.writerow(latlong)
        finally:
            _out.close()
    finally:
        _in.close()


def parse_arguments():
    parser = ap.ArgumentParser(description='Converts UTM coordinates into latitude/longitude.')

    parser.add_argument('-i', '--input_data_file',
                        help='Step file containing input data to manipulate.')
    parser.add_argument('-o', '--output_data_file', help='Step file to store output data.')
    parser.add_argument('-l', '--log_file', help='Step file to collect log information.')

    return parser.parse_args()


if __name__ == '__main__':
    args = parse_arguments()

    utm_to_latlong(args.input_data_file, args.output_data_file, args.log_file)
