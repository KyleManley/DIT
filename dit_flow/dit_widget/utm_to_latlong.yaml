type: manipulation_widget
description: Converts UTM coordinates into latitude/longitude.
