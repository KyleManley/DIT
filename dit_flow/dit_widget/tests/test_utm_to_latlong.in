621988,6088495,19, True
621900,6089268,19, True
621534,6089464,19, True
621630,6089558,19, True
