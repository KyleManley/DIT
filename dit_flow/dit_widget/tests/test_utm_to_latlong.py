import os

from dit_flow.dit_widget.utm_to_latlong import utm_to_latlong


def test_utm_to_latlong(tmpdir):
    this_dir = os.path.dirname(os.path.realpath(__file__))
    in_data_file = os.path.join(this_dir, 'test_utm_to_latlong.in')
    temp_out_file = tmpdir.mkdir("out").join('output_file.csv')
    temp_log_file = tmpdir.mkdir("log").join('log_file.txt')
    utm_to_latlong(input_data_file=in_data_file,
                   output_data_file=temp_out_file.strpath,
                   log_file=temp_log_file)
    actual_out = temp_out_file.read()
    actual_log = temp_log_file.read()
    print(actual_out)
    print(actual_log)
    expected_out_file = os.path.join(this_dir, 'test_utm_to_latlong.latlong.expected')
    expected_log_file = os.path.join(this_dir, 'test_utm_to_latlong.log.expected')
    with open(expected_out_file) as open_out, open(expected_log_file) as open_log:
        expected_out = open_out.read()
        expected_log = open_log.read()

    assert expected_out in actual_out

    for expected_line in expected_log:
        assert expected_line in actual_log
