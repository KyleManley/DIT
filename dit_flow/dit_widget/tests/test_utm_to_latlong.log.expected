Changed row 0 from: ('621988', '6088495')  to: (-35.3393644819495, -67.65764427168914)
Changed row 1 from: ('621900', '6089268')  to: (-35.33240705055096, -67.65872754754074)
Changed row 2 from: ('621534', '6089464')  to: (-35.33068481959214, -67.66278279394442)
Changed row 3 from: ('621630', '6089558')  to: (-35.3298257715448, -67.661740748294)
